root/static/scripts/edit/components/ArtistCreditBubble.js

```diff
@@ -1,172 +1,304 @@
-    <div className="buttons">
-      <button
-        onClick={copyArtistCredit}
-        style={{float: 'left'}}
-        type="button"
-      >
-        {l('Copy Credits')}
-      </button>
-      <button
-        onClick={pasteArtistCredit}
-        style={{float: 'left'}}
-        type="button"
-      >
-        {l('Paste Credits')}
-      </button>
+      ) : null}
+      <tr className="artist-credit-header">
+        <th style={WIDTH_40PC}>{l('Artist in MusicBrainz:')}</th>
+        <th style={WIDTH_40PC}>{l('Artist as credited:')}</th>
+        <th>{l('Join phrase:')}</th>
+        <th />
+      </tr>
+    </thead>
+  );
+}));
+
+type AddArtistCreditRowPropsT = {
+  +dispatch: (ActionT) => void,
+};
+
+const AddArtistCreditRow = (React.memo<
+  AddArtistCreditRowPropsT,
+>(({
+  dispatch,
+}) => (
+  <tr>
+    <td className="align-right" colSpan="4">
       <button
-        className="positive"
-        onClick={done}
-        style={{float: 'right'}}
+        className="add-item with-label"
+        onClick={() => dispatch({type: 'add-name'})}
         type="button"
       >
-        {l('Done')}
+        {l('Add Artist Credit')}
       </button>
-      {extraButtons ? extraButtons : null}
-    </div>
+    </td>
+  </tr>
+)));
+
+type ChangeMatchingTrackArtistsRowPropsT = {
+  +changeMatchingTrackArtists: boolean | void,
+  +dispatch: (ActionT) => void,
+  +initialArtistCreditString: string,
+};
+
+const ChangeMatchingTrackArtistsRow = (React.memo<
+  ChangeMatchingTrackArtistsRowPropsT,
+>(({
+  changeMatchingTrackArtists,
+  dispatch,
+  initialArtistCreditString,
+}) => (
+  <div>
+    <label>
+      <input
+        checked={changeMatchingTrackArtists === true}
+        id="change-matching-artists"
+        onChange={
+          (event: SyntheticEvent<HTMLInputElement>) => dispatch({
+            checked: event.currentTarget.checked,
+            type: 'set-change-matching-artists',
+          })
+        }
+        type="checkbox"
+      />
+      {initialArtistCreditString ? (
+        texp.l(
+          'Change all artists on this release that match “{name}”',
+          {name: initialArtistCreditString},
+        )
+      ) : (
+        l('Change all artists on this release that are currently empty')
+      )}
+    </label>
   </div>
-);
+)));
+
+type ArtistCreditBubblePropsT = {
+  +closeAndReturnFocus: () => void,
+  +dispatch: (ActionT) => void,
+  +state: ArtistCreditStateT,
+};
+
+const ArtistCreditBubble = (React.memo<ArtistCreditBubblePropsT>(({
+  closeAndReturnFocus,
+  dispatch,
+  state,
+}: ArtistCreditBubblePropsT): React.MixedElement => {
+  const {
+    changeMatchingTrackArtists,
+    editsPending,
+    entity,
+    initialArtistCreditString,
+    names,
+  } = state;
+
+  const isTrack = entity.entityType === 'track';
+
+  const tableRef = React.useRef<HTMLTableElement | null>(null);
+
+  React.useEffect(() => {
+    const handleEnter = () => {
+      if (
+        isTrack &&
+        // $FlowIgnore[prop-missing]
```

Comment:
When is this missing? When there's no next track? If so, cannot we just check whether the prop exists at all?

Reply:
Well, `entity` is an `ArtistCreditableT` which may or may not be a track in the release editor. Checking `isTrack` just doesn't tell Flow that it's a track. I'll try to improve it.